### DISCUSSION TITLE
Allow injecting queryCache in hooks

### DISCUSSION THREAD

**Discussion by antonobergklarna** [TanStack/query#1353] [Ideas]
Hi there!

We have a large app with 50 or so different teams working with and owning different features. Features contains both routes and in some cases components or services (data api) that is exposed and consumed by other features/teams. Its important for us that these features are isolated and can be loaded lazy. They also have different security requirements and we need to be able to flush state in individual features differently depending on auth levels etc. Features also need to be able to act on events from fetching life-cycles that are initiated from the specific feature. 

We are evaluating react-query and in general it seams like a good fit however we would like to be able to have multiple query caches and inject them into the hooks. We can achieve a similar result by wrapping all hooks and prefixing query-keys with namespaces for features and security-zones and use the prefixes to route events to the correct feature but it would be easier if each feature can have its own cache.  

Note that since we in some cases export methods and hooks that is consumed by other features / other routes we cant use the Provider as a way to use different caches. 

Would it be possible to add a queryCache as an optional param to the hooks? 

This might be related to this discussion;:
https://github.com/tannerlinsley/react-query/discussions/1220

**Comment by TkDodo** [comment]
> Note that since we in some cases export components or methods that is consumed by other features / other routes we can use the Provider as a way to use different caches.

I think you can just get the queryCache with `useQueryCache`. This is just react context, so it will look up the nearest ReactQueryCacheProvider. So you can just use a different provider for parts of the „app“. Can you explain why this is not sufficient?

**Reply by antonobergklarna** [reply]
Thanks for your comment @TkDodo!

My bad type - can = cant and should have mentioned hooks there...

Let me try to clarify with an example:

Let's say that you have a todo feature that is responsible for todos. It does some data-fetching and has routes to edit and view todos. It also exposes some components, functions and hooks that other features can access through the todo feature index file.

If this todo feature has its own queryCache its fine to export components (wrap them with the provider), functions like getters and setters (they could easily access the local queryCache) however it does not work for the hooks. 

Since these exported hooks can be used in other parts of the app than the todo routes we cant rely on having the ReactQueryCacheProvider with the todo queryCache in parent components. The todo feature could expose e.g. a HoC "withTodoQueryCache" that consuming features could use and wrap its component that is using the exposed todo hook. However this is not a very nice developer experience. Especially if a feature exposes a hook that depends on a hook from another feature. Then you would have to export a composed HoC that adds both features queryCaches to context.

**Reply by TkDodo** [reply]
> Since these exported hooks can be used in other parts of the app than the todo routes we cant rely on having the ReactQueryCacheProvider with the todo queryCache in parent components.

Are you sharing those hooks over multiple applications? Because we are generally in the same boat - we have multiple features, e.g.  `todos`, `widgets`, `profile`. The data fetching is usually kept locally to the components, so I haven't had the need to export the `useTodosQuery` hook, as only the todos feature needs it. But if you do, and you want the profile page to also load todos, it would still just work as long as you have on QueryCacheProvider at the top of your "app":

```
import { QueryCache, ReactQueryCacheProvider } from 'react-query'

const queryCache = new QueryCache()

const App = () => (
    <ReactQueryCacheProvider queryCache={queryCache} />
        <Todos />
        <Widgets />
        </Profile />
    </ReactQueryCacheProvider>
)
```

Does every feature need its own cache? As long as they have distinct query keys (one always starts with `['todos']` while the other always starts with `['widgets']`), I think they can just "the one and only" cache. At least that's how we are doing it, and I think this is also recommended. Now, the `Todos` feature can just import `useWidgetsQuery`, and it will works just fine...

**Reply by antonobergklarna** [reply]
We can work with prefix and if we rely on conventions it is ofc quite easy. However we are a lot of teams and history has shown that it is better if conventions can be enforced. 
To enforce this prefix we could create a layer around react-query but intercepting all hooks and all queryCache methods will be some work - especially if we want to support the different ways of interacting with the hooks. 

It might be recommended to use a single cache but as far as I know its not an as strong recommendation as for e.g. redux where they strongly advice against multiple stores. I believe that our use case is not that uncommon in larger apps and I think that allowing injecting the queryCache would open up a lot of opportunities when it comes to creating abstractions. Either internal or open source. 

Also from browsing the code it looks like it should be fairly straight forward to implement.

**Reply by TkDodo** [reply]
Do you have a code example of what you have in mind?

FWIW, you can always export a created cache as well and access it anywhere or pass it as a prop to hooks.

**Reply by antonobergklarna** [reply]
Are you referring to a code example of how I think we need to wrap react-query or how we want to use react-query if its possible to inject the cache in the hooks? In that case no not at the moment but I can create an example if it helps the discussion. 

To me it looks like I cant pass the cache as a prop to the hooks because its overwritten by the cache provided through context:
https://github.com/tannerlinsley/react-query/blob/master/src/core/config.ts#L115

**Reply by TkDodo** [reply]
> To me it looks like I cant pass the cache as a prop to the hooks because its overwritten by the cache provided through context

Yes, it won’t work automatically. I’m still struggling to understand what you want to do, sorry. That’s why i think an example would help. Do you need the queries of feature A to use options of the cache from feature B? Or read/write to that cache? Or something else entirely?

**Reply by antonobergklarna** [reply]
Sorry for the silence. I will come back with an example eventually. A lot to do now before xmas
